Refactor status cmd

hzhou97 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Refactor the `status certificate` command to build a `CertificateStatus` struct, then call `String()` for output. This design is supposed to be more reusable, as we now also have a struct to pass around rather then just string being printed line by line. Should not change any visible behaviour.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```
NONE
```

/kind cleanup
/area ctl

hzhou97 commented:
/assign @meyskens

hzhou97 commented:
/retest

james-w commented on cmd/ctl/pkg/status/certificate/types.go:
```diff
@@ -0,0 +1,410 @@
+}
+
+type CertificateStatusBuilder struct {
+	// Name of the Certificate resource
+	Name string
+	// Namespace of the Certificate resource
+	Namespace string
+	// Creation Time of Certificate resource
+	CreationTime metav1.Time
+	// Conditions of Certificate resource
+	Conditions []cmapiv1alpha2.CertificateCondition
+	// DNS Names of Certificate resource
+	DNSNames []string
+	// Events of Certificate resource
+	Events *v1.EventList
+	// Not Before of Certificate resource
+	NotBefore *metav1.Time
+	// Not After of Certificate resource
+	NotAfter *metav1.Time
+	// Renewal Time of Certificate resource
+	RenewalTime *metav1.Time
```
This is a really good change, but I find this bit strange.

We are storing almost a duplicate of `CertificateStatus` in the builder, when all
the other resources create a `Status` immediately.

I can think of a couple of ways of doing this that avoids all of this duplication:

1) immediately start filling in the `CertificateStatus` when calling the builder, rather than storing a copy and then copying in `.build()`.

2) Just have a `.Certificate` field on the builder, and then copy the info to the status in `.build()` rather than storing
the intermediate copy.

What do you think?

hzhou97 commented:
I changed it to populating the fields of `CertificateStatus` directly and deleted `CertificateStatusBuilder`
cc @james-w @wallrj 

Also @wallrj re empty fields, I think that a field being empty can be useful information as well? So the user can see what is empty but shouldn't be instead of not knowing what is missing when we omit it altogether.

james-w commented:
I think this looks good. I agree that the empty fields could be useful, but I don't have strong opinion either way.

I'm happy if Richard is happy :)

hzhou97 commented:
/assign @meyskens 

Approval needed

meyskens commented:
/approve
